WalletWasabi.Fluent/ViewModels/CoinSelection/SelectCoinsDialogViewModel.cs

```diff
@@ -0,0 +1,183 @@
+
+		IsSelectionBadlyChosen = allCoins.Select(coins => IsSelectionBadForPrivacy(coins, _transactionLabels)).ReplayLastActive();
+
+		SelectedAmount = selectedCoins.Select(Sum);
+
+		RemainingAmount = SelectedAmount.Select(money => Money.Max(TargetAmount - money, Money.Zero));
+
+		SelectedCount = selectedCoins.Select(models => models.Count);
+
+		NextCommand = ReactiveCommand.CreateFromObservable(() => selectedCoins, EnoughSelected);
+		NextCommand.Subscribe(models => Close(DialogResultKind.Normal, GetCoins().Where(x => models.Any(coin => coin.OutPoint == x.OutPoint))));
+
+		SelectPredefinedCoinsCommand = ReactiveCommand.Create(() => coinCollection.ToList().ForEach(x => x.IsSelected = _usedCoins.Any(coin => x.Coin.OutPoint == coin.OutPoint)));
+
+		SelectAllCoinsCommand = ReactiveCommand.Create(() => coinCollection.ToList().ForEach(x => x.IsSelected = true));
+
+		ClearCoinSelectionCommand = ReactiveCommand.Create(() => coinCollection.ToList().ForEach(x => x.IsSelected = false));
+
+		SelectAllPrivateCoinsCommand = ReactiveCommand.Create(() => coinCollection.ToList().ForEach(coinViewModel => coinViewModel.IsSelected = coinViewModel.PrivacyLevel == PrivacyLevel.Private));
+
+		var commands = new[]
+		{
+			new CommandViewModel("All", SelectAllCoinsCommand),
+			new CommandViewModel("None", ClearCoinSelectionCommand),
+			new CommandViewModel("Private coins", SelectAllPrivateCoinsCommand),
+			new CommandViewModel("Smart", SelectPredefinedCoinsCommand)
+		};
+
+		CoinBasedSelection = new CoinBasedSelectionViewModel(coinChanges, commands)
+			.DisposeWith(disposables);
+
+		LabelBasedSelection = new LabelBasedCoinSelectionViewModel(coinChanges, commands)
+			.DisposeWith(disposables);
+
+		SelectPredefinedCoinsCommand.Execute()
+			.Subscribe()
+			.DisposeWith(disposables);
+
+		base.OnNavigatedTo(isInHistory, disposables);
+	}
+
+	private static Money Sum(IEnumerable<SelectableCoin> coinViewModels)
+	{
+		return coinViewModels.Sum(coinViewModel => coinViewModel.Amount);
+	}
+
+	private bool IsSelectionBadForPrivacy(IReadOnlyCollection<SelectableCoin> coins, SmartLabel transactionLabel)
+	{
+		var selectedCoins = coins.Where(x => x.IsSelected).ToList();
+		var privateAndSemiPrivate = coins.Where(x => x.PrivacyLevel == PrivacyLevel.Private || x.PrivacyLevel == PrivacyLevel.SemiPrivate).ToList();
+		var privateCoinsAreEnough = privateAndSemiPrivate.Sum(x => x.Amount) >= TargetAmount;
+		var isNonPrivateSelected = selectedCoins.Any(x => x.PrivacyLevel != PrivacyLevel.Private);
+
+		if (privateCoinsAreEnough && isNonPrivateSelected)
+		{
+			return true;
+		}
+
+		if (selectedCoins.All(x => x.SmartLabel == transactionLabel))
+		{
+			return false;
+		}
+
+		if (selectedCoins.Any(x => x.AnonymitySet == 1))
+		{
+			return true;
+		}
+
+		if (selectedCoins.GroupBy(x => x.PrivacyLevel).Count() > 1)
+		{
+			return true;
+		}
+
+		if (selectedCoins.GroupBy(x => x.SmartLabel).Count() > 1)
+		{
+			return true;
+		}
+
+		return false;
+	}
```

Comment:
The logic here is a bit messy.
We can simplify it to warn only when a coin with anonymity set 1 is used, except when using only coins with the same label as the recipient or combining that with private or semi-private coins.

```suggestion
	private bool IsSelectionBadForPrivacy(IReadOnlyCollection<SelectableCoin> coins, SmartLabel transactionLabel)
	{
		var selectedCoins = coins.Where(x => x.IsSelected).ToList();

		if (selectedCoins.All(x => x.SmartLabel == transactionLabel || x.PrivacyLevel == PrivacyLevel.Private || x.PrivacyLevel == PrivacyLevel.SemiPrivate))
		{
			return false;
		}

		if (selectedCoins.Any(x => x.AnonymitySet == 1))
		{
			return true;
		}

		return false;
	}
```

Reply:
@SuperJMN I merged this one in order to easily continue reviewing and testing this.